scripts/presets/travis/sharded_clusters/cluster.json

```diff
@@ -69,6 +69,13 @@
             "logappend": true,
             "port": 4300,
             "bind_ip_all": true
+        },
```

Comment:
This change and all other changes in this commit are necessary because a session is not pinned when the topology contains only a single mongos node. I'm not sure if that's a bug in libmongoc, as the spec doesn't specify whether a session needs to be pinned when only connected to a single mongos. After all, that also ensures that all commands are sent to the same server during the transaction.

Reply:
libmongoc considers a URI like "mongodb://foo.example.com/" to be topology type `Single`. We only pin for topology type `Sharded`:
https://github.com/mongodb/mongo-c-driver/blob/10f6ee9458449790125410a6293433a39aa9e9c1/src/libmongoc/src/mongoc/mongoc-cluster.c#L510-L512

I'm not sure if there's harm in not pinning when there's only one mongos, but I agree we should clarify in the spec.

SRV polling will only applies to topology type `Sharded` or `Unknown` (see [here](https://github.com/mongodb/mongo-c-driver/blob/59742684040c89708bfe7ed857efb38c65f30d98/src/libmongoc/src/mongoc/mongoc-topology.c#L486-L491)), so I think we might be safe from SRV polling adding more hosts in the middle of a sharded transaction. The [SRV polling spec](https://github.com/mongodb/specifications/blob/master/source/polling-srv-records-for-mongos-discovery/polling-srv-records-for-mongos-discovery.rst#abstract) says:

> This feature is only available when the Server Discovery has determined that the TopologyType is Sharded, or Unknown. 

And based on:
https://github.com/mongodb/mongo-c-driver/blob/59742684040c89708bfe7ed857efb38c65f30d98/src/libmongoc/src/mongoc/mongoc-topology.c#L304-L344

If SRV lookup only returned one host, I think we would transition to topology type `Single`. That seems to abide correctly by the [Initial DNS Seedlist Discovery Spec](https://github.com/mongodb/specifications/blob/master/source/initial-dns-seedlist-discovery/initial-dns-seedlist-discovery.rst):
> Client MUST treat the host information as if the following URI was used
(i.e. we construct a URI from the records returned and proceed as normal, transitioning to `Single` as if this URI was passed directly).

So, if we used a `mongodb+srv` URI, which initially resolved to a single host like:
`mongodb://host1`, we would transition to topology type `Single` and never scan. And IIUC if we start as topology type `Sharded` and in the process of polling remove all hosts but one, we will *not* transition back to `Single`.

Reply:
> IIUC if we start as topology type Sharded and in the process of polling remove all hosts but one, we will not transition back to Single.

In this case, since the topology would remain sharded we could trust that libmongoc would still pin the session even in the case where there was only a single mongos in the topology. Is that correct? If so, I think @alcaeus' original change in this PR is correct and we should just ensure that all of our sharded cluster MO configurations define multiple mongos hosts.

Reply:
> Is that correct?

I think so, since we don't have active tests, I ran through a quick manual test (following instructions I documented when reviewing SRV polling [here](https://github.com/kevinAlbs/SRVPollingTesting)). Here's some log output of a test where the initial seedlist has two hosts, then drops to one host:

```
it has been 6 seconds
host_list:
-mongodb1.mongodb.com:27018
-mongodb2.mongodb.com:27017
topology type: Sharded
it has been 7 seconds
host_list:
-mongodb1.mongodb.com:27018
-mongodb2.mongodb.com:27017
topology type: Sharded
it has been 8 seconds
host_list:
-mongodb1.mongodb.com:27018
topology type: Sharded
it has been 9 seconds
host_list:
-mongodb1.mongodb.com:27018
topology type: Sharded
it has been 10 seconds
```

The topology type remains sharded.

Somewhat unrelated, but if the initial seedlist only resolves to one host, we do not poll. I'm not sure if that was intentional in the SRV polling spec. I created https://jira.mongodb.org/browse/SPEC-1429.

Reply:
> libmongoc considers a URI like "mongodb://foo.example.com/" to be topology type `Single`.

That explains why the PHP driver tests failed with a single mongos. Since it seems impossible for now that we start out with a single mongos and then discover more of them, I'm perfectly fine leaving things in the C driver as is, barring changes from SPEC-1429. This will not require any changes to the PHP driver directly, since we rely on libmongoc to pin the session to a selected server if necessary. I've closed [CDRIVER-3333](https://jira.mongodb.org/browse/CDRIVER-3333) as "Works as Designed".